src/rust/bitbox02/src/ui.rs

```diff
@@ -54,6 +54,68 @@ pub fn trinary_input_string_create_password(
     Component(component)
 }
 
+#[derive(Default)]
+pub struct ConfirmParams<'a> {
+    /// The confirmation title of the screen. Max 200 chars, otherwise **panic**.
+    pub title: &'a str,
+    /// The confirmation body of the screen. Max 200 chars, otherwise **panic**.
+    pub body: &'a str,
+    /// If true, the body is horizontally scrollable.
+    pub scrollable: bool,
+    /// If true, require the hold gesture to confirm instead of tap.
+    pub longtouch: bool,
+    /// If true, the user can only confirm, not reject.
+    pub accept_only: bool,
+    /// if true, the accept icon is a right arrow instead of a checkmark (indicating going to the
+    /// "next" screen).
+    pub accept_is_nextarrow: bool,
+    /// if true, will only show first and last 32 bytes.
+    pub shorten_body: bool,
+    /// Print the value of this variable in the corner. Will not print when 0
+    pub display_size: usize,
+}
+
+/// Creates a user confirmation dialog screen.
+/// `result` - will be asynchronously set to `Some(bool)` once the user accets or rejects.
+pub fn confirm_create(params: &ConfirmParams, result: Pin<&mut Option<bool>>) -> Component {
```

Comment:
It feels like you shouldn't need `Pin` here. Why are you using that?

Reply:
The pointer is sent into C land, so it has to stay constant. With some permutation of the code I had hardfaults without pinning.

Reply:
sounds like a red herring. in case you put it on the heap with box the position should stay the same. if you move the box only the pointer is moved. not the content

Reply:
Note that confirm_create does not say that the result has to a Box, just a Pin - the caller could have a different pin (I wanted to try stack-pinned).

Ignoring that:

Is there a guarantee that heap objects can't be moved? As far as I can tell there isn't, which is why Box::pin exists. Am I misunderstanding it?

Reading the doc for Box::pin:

> Constructs a new Pin<Box<T>>. If T does not implement Unpin, then x will be pinned in memory and unable to be moved.

Sounds like I also have to somehow make sure Unpin is not implemented? :thinking:

Reply:
Possibly relevant:

https://www.reddit.com/r/rust/comments/b9sydt/trying_to_understand_how_rust_juggles_memoryraw/ek6nu89/

Still trying to wrap my head around it though.

Reply:
Pinning is meant for types that contains references to itself, e.g.:

```rust
struct A {
    bu:f [u8;150];
    writer_pos: &'self [u8]
}
```

I've made up the `'self` lifetime. Implementing this type requires that the type doesn't move ever. And if you have a pinned pointer to it, there is a guarantee that it didn't move between operations. You can always create a pinned variable, but you can only unpin types that are "Unpin" because they don't have any of these pesky self-references.

Reply:
I think the problems you are experiencing is that the type gets dropped while you are still in C-land. I'm not sure exactly why though. It looks like you are doing `into_raw()` to consume the box, and then `from_raw` to construct it again.

Reply:
What about this:

```Rust
use std::marker::PhantomPinned;

#[derive(Debug)]
struct NotUnpinnable(Option<i32>, PhantomPinned);

fn main() {
    // This program compiles and is wrong. Changing Box::new to Box::pin fixes it, as it does not
    // let you compile it.
    let mut a = Box::new(NotUnpinnable(Some(1), PhantomPinned));
    let mut b = Box::new(NotUnpinnable(Some(2), PhantomPinned));

    let pointer_to_a: *mut NotUnpinnable = &mut *a;

    let mut c = a; // move a
    // A dangerous operation prevented by pinning
    std::mem::swap(&mut c, &mut b);

    unsafe {
        // would happen in C.
        *pointer_to_a = NotUnpinnable(None, PhantomPinned);
    }

    // Expect `None`, but contains `Some(2)` due to mem::swap.
    println!("{:#?}", c);
}
```

Reply:
> `confirm_created`

I don't want to reimplement each component in Rust now. The goal is async workflows, not porting/rewriting everything that it depends on. trinary input string for example would be a lot harder to port.

Reply:
Pretty sure that without unpin, the pointer can become dangling when the value is accidentally moved, and proper pinning prevents that. Self-referential structs are not the only use case.

Could be wrong though, kind of hard to wade through this :)

Reply:
after long offline discussions, tldr: Pin+!Unpin can prevent data moves, but it can't prevent a premature `drop()` of the result before it is written to asynchronously by C.

Pin is also not needed when using Box::into_raw/from_raw to pass ownership to C. 

As a result, we have to move the workflow/* code to the unsafe component creation, where we have control over everything and can promise the result is not moved/dropped.